presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementMultipleDistinctAggregation.java

```diff
@@ -0,0 +1,140 @@
+ * </pre>
+ * into
+ * <pre>
+ * - Aggregation
+ *        F1(...) mask ($0)
+ *        F2(...) mask ($1)
+ *     - MarkDistinct (k, s0, s1, ...) -> $0
+ *          - MarkDistinct (k, s2, s3, ...) -> $1
+ *              - X
+ * </pre>
+ */
+public class ImplementMultipleDistinctAggregation
+        implements Rule<AggregationNode>
+{
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(ImplementMultipleDistinctAggregation::hasDistinct);
+
+    private static boolean hasDistinct(AggregationNode aggregation)
+    {
+        return aggregation.getAggregations()
+                .values().stream()
+                .anyMatch(e -> e.getCall().isDistinct() && !e.getMask().isPresent());  // MarkDistinct doesn't support a mask (yet), so skip if aggregation has mask set
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode parent, Captures captures, Context context)
+    {
+        // the distinct marker for the given set of input columns
+        Map<Set<Symbol>, Symbol> markers = new HashMap<>();
+
+        Map<Symbol, Aggregation> aggregations = new HashMap<>();
+        PlanNode subPlan = parent.getSource();
+
+        for (Map.Entry<Symbol, Aggregation> entry : parent.getAggregations().entrySet()) {
+            Aggregation aggregation = entry.getValue();
+            FunctionCall call = aggregation.getCall();
+
+            if (call.isDistinct()) {
+                Set<Symbol> inputs = call.getArguments().stream()
+                        .map(Symbol::from)
+                        .collect(Collectors.toSet());
+
+                Symbol marker = markers.get(inputs);
+                if (marker == null) {
+                    marker = context.getSymbolAllocator().newSymbol(inputs.iterator().next().getName(), BOOLEAN, "distinct");
+                    markers.put(inputs, marker);
+                }
+
+                ImmutableSet.Builder<Symbol> distinctSymbols = ImmutableSet.<Symbol>builder()
+                        .addAll(parent.getGroupingKeys())
+                        .addAll(inputs);
+
+                parent.getGroupIdSymbol().ifPresent(distinctSymbols::add);
+
+                // remove the distinct flag and set the distinct marker
+                aggregations.put(entry.getKey(),
+                        new Aggregation(
+                                new FunctionCall(
+                                        call.getName(),
+                                        call.getWindow(),
+                                        call.getFilter(),
+                                        call.getOrderBy(),
+                                        false,
+                                        call.getArguments()),
+                                aggregation.getSignature(),
+                                Optional.of(marker)));
+
+                subPlan = new MarkDistinctNode(
```

Comment:
+1. You could also build `MarkDistinctNodes` from `Map<Set<Symbol>, Symbol> markers = new HashMap<>();` after you pass over all aggregations.

> Alternatively, maybe we could create it always (and maybe not even cache markers here). It could be solved by another rule which would find a cascade of MarkDistinctNodes which have the same distinctSymbols so they could be optimized.

Good idea. They don't also have to have the same `distinctSymbols` but rather widening `distinctSymbols`, e.g: `[a] -> [a, b] -> [a, b, c]`...
It seems that finding optimal order might be cost decision. Is it NP hard problem?
Consider symbol sets: `[a], [a, b], [a, c]`. You have two options:
```
1. [a], [a,b], [a, c]
2. [a], [a,c], [a, b]
```

Reply:
Oops.. That’s what I meant to do. In a previous version it was iterating over the markers instead of the aggregations